Libraries/LLM/Llama.swift

```diff
@@ -7,6 +7,86 @@ import MLXNN
 
 // port of https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/models/llama.py
 
+private class DynamicNTKScalingRoPE: Module {
+    let dims: Int
+    let maxPositionEmbeddings: Int
+    let traditional: Bool
+    let originalBase: Float
+    var scale: Float
+    let ropeType: String
+    let ropeScaling: [String: StringOrNumber]?
+
+    init(
+        dims: Int, maxPositionEmbeddings: Int = 2048, traditional: Bool = false,
+        base: Float = 10000, scale: Float = 1.0, ropeType: String = "default",
+        ropeScaling: [String: StringOrNumber]? = nil
+    ) {
+        self.dims = dims
+        self.maxPositionEmbeddings = maxPositionEmbeddings
+        self.traditional = traditional
+        self.originalBase = base
+        self.scale = scale
+        self.ropeType = ropeType
+        self.ropeScaling = ropeScaling
+    }
+
+    func computeBaseFrequency() -> Float {
+        switch ropeType {
+        case "llama3":
+            return computeLlama3BaseFrequency()
+        default:
+            return originalBase
+        }
+    }
+
+    func computeLlama3BaseFrequency() -> Float {
+        guard let ropeScaling = ropeScaling else {
+            return originalBase
+        }
+
+        guard case .float(let factor) = ropeScaling["factor"],
+            case .float(let lowFreqFactor) = ropeScaling["low_freq_factor"] ?? .float(1.0),
+            case .float(let highFreqFactor) = ropeScaling["high_freq_factor"] ?? .float(4.0),
+            case .float(let oldContextLen) = ropeScaling["original_max_position_embeddings"]
+                ?? .float(8192)
+        else {
+            return originalBase
+        }
+
+        let lowFreqWavelen = oldContextLen / lowFreqFactor
+        let highFreqWavelen = oldContextLen / highFreqFactor
+
+        let freqs = (0 ..< dims).compactMap { index -> Float? in
+            if index % 2 == 0 {
+                return pow(originalBase, Float(index) / Float(dims))
+            }
+            return nil
+        }
+
+        let newBaseFreqs = freqs.map { freq -> Float in
+            let wavelen = 2 * .pi / freq
+            let smooth = max(
+                0, min(1, (wavelen - highFreqWavelen) / (lowFreqWavelen - highFreqWavelen)))
+            return freq * ((1 - smooth) * factor + smooth)
+        }
+
+        return newBaseFreqs.reduce(0, +) / Float(newBaseFreqs.count)
+    }
+
+    func callAsFunction(_ x: MLXArray, offset: Int = 0) -> MLXArray {
+        let seqLen = x.dim(1) + offset
+        var base = computeBaseFrequency()
+        if maxPositionEmbeddings > 0 && seqLen > maxPositionEmbeddings {
```

Comment:
Probably it should be optional and we should do something like `maxPositionEmbeddings != nil && seqLen > maxPositionEmbeddings!`